File: class25/java/GQS2501-SimulacaoPipeline.java
/**
 * GQS2501 - Simulação de Pipeline de CI/CD (build -> test -> analyse -> deploy).
 * Executa em sequência as etapas típicas de um pipeline (Slide 7 da Aula 25),
 * imprimindo o progresso de cada etapa e aplicando a regra de "fail fast": se uma
 * etapa falhar, as etapas seguintes NÃO são executadas e o pipeline é marcado
 * como reprovado -- sem depender de Jenkins, GitLab CI ou GitHub Actions reais.
 *
 * Compilação e execução:
 *   javac GQS2501-SimulacaoPipeline.java
 *   java -cp . GQS2501_SimulacaoPipeline
 */
import java.util.*;

class Etapa {
    String nome;
    double duracaoS;
    double chanceSucesso;
    boolean sucesso;
    String mensagem = "";

    Etapa(String nome, double duracaoS, double chanceSucesso) {
        this.nome = nome;
        this.duracaoS = duracaoS;
        this.chanceSucesso = chanceSucesso;
    }

    boolean executar(Random rng) {
        try {
            Thread.sleep((long) (Math.min(duracaoS, 0.05) * 1000));
        } catch (InterruptedException ignored) {
        }
        sucesso = rng.nextDouble() < chanceSucesso;
        mensagem = sucesso ? "OK" : "FALHOU";
        return sucesso;
    }
}

class Pipeline {
    String nome;
    List<Etapa> etapas;
    Random rng;
    String etapaQueFalhou = null;

    Pipeline(String nome, List<Etapa> etapas, long seed) {
        this.nome = nome;
        this.etapas = etapas;
        this.rng = new Random(seed);
    }

    boolean executar() {
        System.out.println("=".repeat(60));
        System.out.printf("  PIPELINE: %s%n", nome);
        System.out.println("=".repeat(60));

        for (int i = 0; i < etapas.size(); i++) {
            Etapa etapa = etapas.get(i);
            System.out.printf("%n[%d/%d] Executando: %s ... ", i + 1, etapas.size(), etapa.nome);
            boolean ok = etapa.executar(rng);
            System.out.printf("[%s] (%.1fs)%n", etapa.mensagem, etapa.duracaoS);
            if (!ok) {
                etapaQueFalhou = etapa.nome;
                System.out.printf("%n>> FAIL FAST: etapa '%s' falhou. Etapas seguintes NÃO serão executadas.%n",
                        etapa.nome);
                return false;
            }
        }
        return true;
    }

    void imprimirResultadoFinal() {
        System.out.println("\n" + "=".repeat(60));
        if (etapaQueFalhou == null) {
            System.out.println("  RESULTADO FINAL: PIPELINE APROVADO -- deploy concluído com sucesso");
        } else {
            System.out.printf("  RESULTADO FINAL: PIPELINE REPROVADO na etapa '%s'%n", etapaQueFalhou);
        }
        System.out.println("=".repeat(60));
    }
}

class GQS2501_SimulacaoPipeline {

    static Pipeline criarPipelineSucesso() {
        List<Etapa> etapas = List.of(
            new Etapa("Checkout", 0.2, 1.0),
            new Etapa("Build", 0.5, 0.98),
            new Etapa("Unit Tests", 0.8, 0.95),
            new Etapa("Static Analysis (Quality Gate)", 0.6, 0.95),
            new Etapa("Package", 0.3, 0.99),
            new Etapa("Deploy Staging", 0.4, 0.97),
            new Etapa("Deploy Prod", 0.4, 0.97)
        );
        return new Pipeline("app-techstore (branch: main)", etapas, 7);
    }

    static Pipeline criarPipelineComFalha() {
        List<Etapa> etapas = List.of(
            new Etapa("Checkout", 0.2, 1.0),
            new Etapa("Build", 0.5, 1.0),
            new Etapa("Unit Tests", 0.8, 0.0),
            new Etapa("Static Analysis (Quality Gate)", 0.6, 1.0),
            new Etapa("Package", 0.3, 1.0),
            new Etapa("Deploy Staging", 0.4, 1.0),
            new Etapa("Deploy Prod", 0.4, 1.0)
        );
        return new Pipeline("app-techstore (branch: feature/checkout-com-bug)", etapas, 1);
    }

    public static void main(String[] args) {
        Pipeline pipeline1 = criarPipelineSucesso();
        pipeline1.executar();
        pipeline1.imprimirResultadoFinal();

        System.out.println();

        Pipeline pipeline2 = criarPipelineComFalha();
        pipeline2.executar();
        pipeline2.imprimirResultadoFinal();
    }
}

File: class25/python/GQS2501-SimulacaoPipeline.py
#!/usr/bin/env python3
"""
GQS2501 - Simulação de Pipeline de CI/CD (build -> test -> analyse -> deploy).
Executa em sequência as etapas típicas de um pipeline (Slide 7 da Aula 25),
imprimindo o progresso de cada etapa e aplicando a regra de "fail fast": se uma
etapa falhar, as etapas seguintes NÃO são executadas e o pipeline é marcado
como reprovado -- sem depender de Jenkins, GitLab CI ou GitHub Actions reais.

Uso: python3 GQS2501-SimulacaoPipeline.py
"""
import random
import time


class Etapa:
    """Representa uma etapa (stage) do pipeline."""

    def __init__(self, nome: str, duracao_s: float, chance_sucesso: float):
        self.nome = nome
        self.duracao_s = duracao_s
        self.chance_sucesso = chance_sucesso
        self.sucesso: bool = False
        self.mensagem: str = ""

    def executar(self, rng: random.Random) -> bool:
        """Simula a execução da etapa (sem realmente esperar o tempo real,
        só uma pequena pausa para efeito didático)."""
        time.sleep(min(self.duracao_s, 0.05))  # pausa curta, só para dar sensação de progresso
        self.sucesso = rng.random() < self.chance_sucesso
        self.mensagem = "OK" if self.sucesso else "FALHOU"
        return self.sucesso


class Pipeline:
    """Executa uma lista de etapas em sequência, com fail fast."""

    def __init__(self, nome: str, etapas: list, seed: int = 7):
        self.nome = nome
        self.etapas = etapas
        self.rng = random.Random(seed)
        self.etapa_que_falhou: str | None = None

    def executar(self) -> bool:
        print("=" * 60)
        print(f"  PIPELINE: {self.nome}")
        print("=" * 60)
        for indice, etapa in enumerate(self.etapas, start=1):
            print(f"\n[{indice}/{len(self.etapas)}] Executando: {etapa.nome} ...", end=" ")
            ok = etapa.executar(self.rng)
            print(f"[{etapa.mensagem}] ({etapa.duracao_s:.1f}s)")
            if not ok:
                self.etapa_que_falhou = etapa.nome
                print(f"\n>> FAIL FAST: etapa '{etapa.nome}' falhou. "
                      f"Etapas seguintes NÃO serão executadas.")
                return False
        return True

    def imprimir_resultado_final(self) -> None:
        print("\n" + "=" * 60)
        if self.etapa_que_falhou is None:
            print("  RESULTADO FINAL: PIPELINE APROVADO -- deploy concluído com sucesso")
        else:
            print(f"  RESULTADO FINAL: PIPELINE REPROVADO na etapa '{self.etapa_que_falhou}'")
        print("=" * 60)


def criar_pipeline_sucesso() -> Pipeline:
    """Pipeline com alta chance de sucesso em todas as etapas (cenário feliz)."""
    etapas = [
        Etapa("Checkout", 0.2, chance_sucesso=1.0),
        Etapa("Build", 0.5, chance_sucesso=0.98),
        Etapa("Unit Tests", 0.8, chance_sucesso=0.95),
        Etapa("Static Analysis (Quality Gate)", 0.6, chance_sucesso=0.95),
        Etapa("Package", 0.3, chance_sucesso=0.99),
        Etapa("Deploy Staging", 0.4, chance_sucesso=0.97),
        Etapa("Deploy Prod", 0.4, chance_sucesso=0.97),
    ]
    return Pipeline("app-techstore (branch: main)", etapas, seed=7)


def criar_pipeline_com_falha() -> Pipeline:
    """Pipeline onde a etapa de testes unitários é forçada a falhar,
    demonstrando o comportamento de fail fast."""
    etapas = [
        Etapa("Checkout", 0.2, chance_sucesso=1.0),
        Etapa("Build", 0.5, chance_sucesso=1.0),
        Etapa("Unit Tests", 0.8, chance_sucesso=0.0),  # forçado a falhar
        Etapa("Static Analysis (Quality Gate)", 0.6, chance_sucesso=1.0),
        Etapa("Package", 0.3, chance_sucesso=1.0),
        Etapa("Deploy Staging", 0.4, chance_sucesso=1.0),
        Etapa("Deploy Prod", 0.4, chance_sucesso=1.0),
    ]
    return Pipeline("app-techstore (branch: feature/checkout-com-bug)", etapas, seed=1)


def main() -> None:
    pipeline1 = criar_pipeline_sucesso()
    pipeline1.executar()
    pipeline1.imprimir_resultado_final()

    print("\n")

    pipeline2 = criar_pipeline_com_falha()
    pipeline2.executar()
    pipeline2.imprimir_resultado_final()


if __name__ == "__main__":
    main()
